Bump gene-expression-volume and dependencies

GianlucaFicarelli opened the pull request:
Bump gene-expression-volume and dependencies.

repo-bluebrain:
- gene-expression-volume 0.1.0 -> 0.1.1
- py-atlalign 0.6.0 -> 0.6.1 (moved from `bluebrain/repo-patches`)
- py-atldld 0.3.2 -> 0.3.3

repo-patches:
- llvm (added `libs` property from upstream)
- py-antspyx 0.2.7 -> 0.3.1
- py-tensorflow-addons 0.13.0 -> 0.16.1
- py-torch (the checksum of the patch changed?)
- py-tensorflow AND tensorflow removed

builtin (from upstream, but with some patches, mainly because of the changes to the PythonPackage class, I'll attach the diff):
- py-keras 2.4.3 -> 2.7.0
- py-libclang 11.1.0 (new)
- py-tensorboard-data-server 0.6.1 (new)
- py-tensorboard-plugin-wit 1.8.0 -> 1.8.1
- py-tensorboard 2.4.1 -> 2.7.0
- py-tensorflow-estimator 2.4.0 -> 2.7.0
- py-tensorflow-probability (fixed tmp directory)
- py-tensorflow 2.5.1 -> 2.7.0

GianlucaFicarelli committed:
Bump gene-expression-volume and dependencies

Dependencies:

- py-atlalign
- py-atldld
- py-tensorflow (from upstream)
- py-tensorflow-addons
- py-libclang (from upstream)
- py-tensorboard-data-server (from upstream)
- py-tensorboard-plugin-wit (from upstream)
- py-tensorboard (from upstream)
- py-tensorflow-estimator (from upstream)

GianlucaFicarelli committed:
Fix licenses

GianlucaFicarelli committed:
Remove tensorflow from repo-patches

matz-e reviewed (commented):
There was some purpose to have our own "patched" tensorflow: to avoid having to build it from source for hours on end. What changes here?

Also, are all changes to the `builtin` repo taken from upstream?

GianlucaFicarelli commented:
> There was some purpose to have our own "patched" tensorflow: to avoid having to build it from source for hours on end. What changes here?

@matz-e Yeah, I wanted to ask, but I'm not on it at the moment (I just pushed some local changes). It would be simpler if we can use the upstream definitions of the packages, but if it takes too long to compile I see the point.

I tried to pick all the changes from upstream, because tensorflow>=2.6.0 is needed for the new version of atlalign, and this version of tensorflow is already available upstream.

Do you think it's simpler to add the new version of tensorflow to our patches?

matz-e commented:
> @matz-e Yeah, I wanted to ask, but I'm not on it at the moment (I just pushed some local changes). It would be simpler if we can use the upstream definitions of the packages, but if it takes too long to compile I see the point.

The simpler our Spack, the better. Let's see how long this takes.

> I tried to pick all the changes from upstream, because tensorflow>=2.6.0 is needed for the new version of atlalign, and this version of tensorflow is already available upstream.

As long as `builtin` contains only stuff backported from upstream, I'm fine. 

> Do you think it's simpler to add the new version of tensorflow to our patches?

I'd actually prefer upstream if it builds. It was a bit of a pain finding out which Python version binaries are provided for :-/

GianlucaFicarelli committed:
Try to remove tensorflow and six pinning

GianlucaFicarelli committed:
Fix wheels build for old version of spack

GianlucaFicarelli committed:
Move py-atlalign to repo-bluebrain

GianlucaFicarelli committed:
Fix for old version of PythonPackage

GianlucaFicarelli committed:
Add py-wheel to py-libclang

GianlucaFicarelli committed:
Fix tmpdir in py-tensor*

GianlucaFicarelli committed:
Sync from upstream

GianlucaFicarelli committed:
Sync llvm dirs and tensorflow python_platlib

GianlucaFicarelli committed:
Add py-antspyx@0.3.1 and py-keras@2.7.0

GianlucaFicarelli committed:
Fix py-tensorflow-addons dependencies

GianlucaFicarelli committed:
Fix py-keras?

GianlucaFicarelli committed:
Fix py-keras, try 2

GianlucaFicarelli commented on bluebrain/repo-bluebrain/packages/py-atlalign/package.py:
```diff
@@ -0,0 +1,32 @@
+    depends_on('python@3.6:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
+
+    depends_on('py-antspyx@0.3.1', type=('build', 'run'))
```
Hi @EmilieDel, I opened this PR to update `gene-expression-volume` and its dependencies in spack, including `atlalign 0.6.1`.
The build is successful now, but I needed to set the `py-antspyx` dependency to the version 0.3.1.

I see that:
- `atlalign` depends on `antspyx==0.2.7` in https://github.com/BlueBrain/atlas-alignment/blob/54b7c2061d94f3f7e8ee27b98f64a024558fa43b/setup.py#L20
- but the source and wheel packages in https://pypi.org/project/atlalign/0.6.1/#files depend on `antspyx==0.3.1`

I wonder if you can help to clarify what is the correct dependency?

As a side note, antspyx causes a lot of warnings to be printed when running gene-expression-volume,  but I'm not sure if we can or want to do something about it.
```
/gpfs/bbp.cscs.ch/ssd/apps/bsd/pulls/1527/stage_applications/install_gcc-9.4.0-skylake/py-antspyx-0.3.1-nofauo/lib/python3.9/site-packages/ants/core/ants_image.py:611: DeprecationWarning:

inspect.getargspec() is deprecated since Python 3.0, use inspect.signature() or inspect.getfullargspec()
```

EmilieDel replied:
Hello @GianlucaFicarelli, sorry about that! We can go with `antspy==0.3.1`. I am going to open a PR on `atlalign` to correct this. 

Thanks for the side note, I am going to have a look at it.

GianlucaFicarelli commented:
@matz-e @pramodk The PR can be reviewed. As expected, building tensorflow requires some time (see the job logs), and it can be compared with the previous builds.

I attached a file with the patches applied to the packages updated in the builtin directory.
Alternatively, the files could be moved to repo-patches, but I'm not sure that it will be easier to update spack in this case.
 
[diff_20220401.patch.txt](https://github.com/BlueBrain/spack/files/8396806/diff_20220401.patch.txt)
[spack_compare.sh.txt](https://github.com/BlueBrain/spack/files/8396804/spack_compare.sh.txt)

(EDIT: updated files)

matz-e commented:
> I attached a file with the patches applied to the packages updated in the builtin directory.
> Alternatively, the files could be moved to repo-patches, but I'm not sure that it will be easier to update spack in this case.

For all modifications to `builtin`: will you lose something if on the next merge we will do
```
git reset --hard upstream/0.18.0 -- var/spack/repos/builtin
```
?

If this resets any work done by us, things have to move to `patches`. If whole packages are moved to `patches`, they will overwrite `builtin` no questions asked, and `patches` is preferred over `builtin`.

Same test, rephrased: did all modifications to `builtin` come from upstream?

matz-e reviewed (commented):
I don't see anything obvious, this all looks fine to me.

Not keen on having yet another LLVM installation times 3, but I think with Python dependencies, these things are unavoidable.

I noticed though that `gene-expression-volume` defaulted to GCC 9.4, investigating.

matz-e committed:
bump to rebuild

GianlucaFicarelli commented:
The PR in upstream that removes the hardcoded /tmp has been already merged, so I'm attaching the new diff.
[diff_20220404.patch.txt](https://github.com/BlueBrain/spack/files/8409214/diff_20220404.patch.txt)

The changes in `var/spack/repos/builtin/packages` can all be overwritten when spack will be updated to the next release (0.18?)

matz-e committed:
attempt to pin compilers

matz-e committed:
can't spell

matz-e committed:
pin tf compiler

matz-e committed:
Pin software directly

matz-e committed:
Add a few externals to chop down the build

matz-e committed:
Add a bazel module

matz-e committed:
generate module for openjdk

matz-e committed:
Thanks, Google!

matz-e committed:
Add patched/butchered bazel in an attempt to catch some of the GCC 11 fixes

matz-e committed:
Bazel henanigans

matz-e committed:
butcher bazel requirement

matz-e committed:
butcher things even more

- bazel@4 is required for tEnsOrFLoW, but doesn't build with gcc@1.11.
  Solution: install a wheel for TF
- TF wants %gcc@9 as it by default relies on an old CUDA version.
  Solution: fix the variant to ~cuda
- Bump bazel to @5 to get stuff working.  TF right now installs a wheel,
  but takes all dependencies from the proper recipe!

matz-e commented:
@GianlucaFicarelli didn't forget about this, I think I finally found the right approach.

Keeping the upstream recipe for all the correct dependencies (assuming that for now until builds are in). TF requires an older bazel that is incapable to build with newer GCC and impossible to patch, as s**t is downloaded and compiled in the configure process. Thus I ended up using the wheel overlaid on top of the build recipe. Old GCC was preferred due to the CUDA dependency.

TensorBoard builds with an updated Bazel, got things to install in my local setup, let's see how the deployment goes. Do we require CUDA for TF?

matz-e committed:
Switch to GPU stuffs.

So great that it just uses the same dir as TF

matz-e committed:
Moar deps

matz-e committed:
Muck around more in this mess

matz-e committed:
pin compiler again

matz-e committed:
bump

matz-e committed:
Sans Bazèle

matz-e committed:
Remove accidental file

matz-e committed:
Fixeth the Toched sha?

matz-e commented:
Seems out of date, I'll close this.
